### DISCUSSION TITLE
0.97.1.4 for MC 1.14.* & 1.15.* & 1.16.* & 1.17.*

### DISCUSSION THREAD

**Discussion by LlmDl** [TownyAdvanced/Towny#5229] [Pre-Release Discussions]
- Bump language file to 0.119.
  - Fix mayors being able to jail their residents in order to get around /t spawn restrictions for enemies and outlaws.
    - Closes #5216.
  - Automatic config update: global_town_settings.prevent_town_spawn_in will have ",outlaw" added to it.
    - Outlaws will be kept from using (and paying for) /t spawn and /n spawn much earlier than before.
  - Fix fishing being able to be bucketed.
    - Closes #5213.
  - Fix /t outlaw add command kicking the command sender from their town.
  - Fix entities not being able to be added to the EventWar explosions_ignore_list.
    - Closes #4791.
  - Fix map hud being off by one townblock in some instances.
  - Fix jailed residents not saving their object immediately.
  - Fix saving of jailed residents in mysql. 
  - Fix /t claim auto|rect not claiming the last townblock in the radius.
    - Closes #5224.
  - Automatic locale detection/Multi-language support, courtesy of Warriorrrr with PR #5167.
    - Your server will now display all towny messages to your players according to their client's locale, if that language is available.
      - There are currently 20 languages available, seen in the config.yml.
      - If your language is not present it can be requested and added to the crowdin website: https://crowdin.com/project/townyadvanced
    - A global.yml language file will be present to allow you to override a language string for all languages.
      - Meaning you do not have to change (for example) the towny prefix in each file, just the global.yml.
    - The console/logs will use the language set in your config.yml.
    - Closes #4926. 
  - Respect the skip named option in onCreatureSpawn, courtesy of Warriorrrr with PR #5227.
    - Closes #5226, Closes #5225.
  
----

<details><summary>Cumulative changes since 0.97.1.0</summary>
<details><summary>0.97.1.3</summary>

  - Placeholder change: %townyadvanced_player_location_pvp% now shows (No PVP) where PVP is disabled.
  - New Placeholder: %townyadvanced_player_location_town_resident_count%
    - If there is a town present where the player stands, it will return the number representing the number of residents that town has.
  - New Placeholder: %townyadvanced_player_location_town_mayor_name%
    - If there is a town present where the player stands, it will return the name of that town's mayor.
  - New Placeholder: %townyadvanced_player_location_town_nation_name%
    - If there is a town present where the player stands, it will return the name of the nation, if the town has one.
  - Language file bumped to 0.118
  - Fix DailyTimerTasks not checking if some lists are empty.
  - Solve potential NPE when residents' jail cells are loaded.
  - Replace deprecated mysql driver.  
  - Fix sending correct helpmessage to ops/admins who log in while safemode is active.
  - Actually fix database.yml not being converted to correctly for servers using mysql. 
</details>
<details><summary>0.97.1.2</summary>

  - Fix database.yml not being converted to the correct loading setting.    
</details>
<details><summary>0.97.1.1</summary>

  - Fix %townyadvanced_player_status% placeholder.
  - New Command: /ta purge [days] {townname}
    - Deletes the players from the given town, who haven't logged in in the given days.
    - Closes #5125.
  - Fix /n set capital not checking the correct number for max-residents-per-town.
  - Fix towns and nations being able to use the economy prefixes as town and nation names.
    - Closes #5196.
  - Update list of banned names to include new /town subcommands.
  - Create database.yml file.
    - Your Towny database settings will be migrated to a new database.yml file.
    - This allows us to be shown the config file without mysql servers having to go to the trouble of removing their passwords.
  - Automatic town and nation name renaming for FlatFile database.
    - If a town or nation would be loaded with an invalid name, which would cause Towny to enter into safemode, Towny will rename that town or nation.
    - Solves for the most common sort of SafeMode in modern Towny.
</details>
</details>

----

<table align=center>
<tr><th colspan=4><h3>Towny Sponsors<br><h4>I want to give a big thank you to all of my sponsors:<br>These are the people that help to make Towny's support and development as active as it is,<br>and who supported me during this pre-release of Towny. </th></tr>
<tr><td>MrLogano</td><td>Momshroom</td><td>Cobrex1</td><td> KarlOfDuty </td></tr>
<tr><td>scatterspell</td><td>gerardvanschip</td><td> DrprofLuigi </td><td>Tobza</td></tr>
<tr><td>SmallSansSerif </td><td>TheNylox</td><td>diontimmer </td><td>trevor1097</td></tr>
<tr><td>benevolent2012</td><td>bretomart</td><td>HunSolos</td><td>0xBit-MC</td></tr>
<tr><td>specs1396</td><td>treeDerevo</td><td>Scottlander100</td><td>ItsSynchro</td></tr>
<tr><td>Advanced-Kind-MC</td><td>MidnightSugar</td><td>Microhazard</td><td>mhiaxo</td></tr>
<tr><td>TheFlagCourier</td><td>PigioGigio</td><td>PestoNotPasta</td><td>catsaretheend</td></tr>
<tr><td>galacticwarrior9</td><td>cigaming</td><td>GreenArrowlol</td><td>KeNickGong</td></tr>
<tr><td>honnisha</td><td>MilesBHuff</td><td>pebblehost</td><td>Akiranya</td></tr>
<tr><td>BamBamTheEggman</td><td>Martajnez</td><td colspan=2>and 2 private sponsors.</td></tr>
<tr><td colspan=4 align=center><h3><a href=https://github.com/sponsors/LlmDl>If you want to support the developer consider becoming a sponsor.</a><br><h6>(It's just like Patreon but 100% goes to the developer.)<br><h3>Plus there are <a href="https://github.com/TownyAdvanced/Towny/wiki/Sponsor-Perks"><b>perks!</b></a></td></tr>
</table>

<table align=center>
<th colspan=3> Important Links </th>
<tr align=center><td><a title="How Towny Works" href="https://github.com/TownyAdvanced/Towny/wiki/How-Towny-Works">How Towny 
Works</a></td><td><a title="Install Guide" href="https://github.com/TownyAdvanced/Towny/wiki/Installation">Towny Install Guide</a></td><td><a title="Updating Towny" href="https://github.com/TownyAdvanced/Towny/wiki/Updating-Towny">Towny Update Guide</a></td>
</tr>
<tr align=center>
<td><a title="Towny Changelog" href="https://raw.githubusercontent.com/TownyAdvanced/Towny/master/resources/ChangeLog.txt" target="_blank" rel="noopener">Complete Changelog</a> </td><td><a title="Default Config Files" href="https://github.com/TownyAdvanced/Towny/wiki/Config-Files">Default Config Files</a></td><td><a title="Commands/Permissions/Placeholders" href="https://github.com/TownyAdvanced/Towny/wiki/Reference">Commands/Permissions<br>Placeholders</a></td>
</tr>
</table>

### ⏬ Download available as a .jar file in the Assets section below:

<hr /><em>This discussion was created from the release <a href='https://github.com/TownyAdvanced/Towny/releases/tag/0.97.1.4'>0.97.1.4 for MC 1.14.* & 1.15.* & 1.16.* & 1.17.*</a>.</em>

**Comment by hmmcrunchy** [comment]
have you had an issue with this error after updating to java 16 

java.lang.RuntimeException: Failed to get driver instance for jdbcUrl=jdbc:mysql://localhost:3306/towny?verifyServerCertificate=false&useSSL=false&useUnicode=true&characterEncoding=utf-8

nothing has changed bar 16 update and towny update, but wont load from data base 

is this something i need to add to java install

**Reply by LlmDl** [reply]
The database settings moved to a new database.yml file, in the towny\settings\ folder. Make sure you have your correct settings in that file, (they should have migrated but did not on a previous pre-release.)

The only other thing which changed regarding the mysql driver was replacing a deprecated version with this: https://github.com/TownyAdvanced/Towny/blob/92c3afabbb0234c41880995e105103aafba4802e/src/com/palmergames/bukkit/towny/db/TownySQLSource.java#L124

**Reply by hmmcrunchy** [reply]
thanks will give it a test, and update to latest again 

the databases file looks ok to me with the same details in 

uninstalling java 8 and installing 16 wouldnt have anything to do with it ?

**Reply by LlmDl** [reply]
I cannot speak to that (re: uninstalling Java 8 for Java 16.) No one else using MC 1.17 and Java 11-16 have had any issues with their mysql connections.

**Reply by hmmcrunchy** [reply]
no worries thanks for the advice ill try and see what's gone wrong

**Reply by hmmcrunchy** [reply]
bizarre rebooted server,  installed java 8 in case, tested database, ran queries in mysql workbench to check running (all ok) - nothing working in spigot just keep getting this message

https://pastebin.com/TL4ePwrj

**Reply by LlmDl** [reply]
Can you give this a shot: https://cdn.discordapp.com/attachments/669247087329673228/880793629717848174/Towny-0.97.1.5.jar

Also, probably not the issue, but can you confirm you have a mysql-connector jar in your server libs folder?
